File: AOC2017/Day2/FSharp/AOC2017-Day2/Program.fs
open System
open System.IO

let IntroFile = @"D:\AdventOfCode\AOC2017\Day2\Inputs\intro.txt";
let ActualFile = @"D:\AdventOfCode\AOC2017\Day2\Inputs\actual.txt"


let ToInt(s: string) =
    s |> int
let ListToInt(s:string[]) =
    s |> Seq.map(ToInt)

let GetMin(n: seq<int>) =
    n |> Seq.min
    
let GetMax(n: seq<int>) =
    n |> Seq.max
 
let ParseLine(line:string) =
    let split = line.Split('\t')
    let nums = ListToInt(split)
    let max = GetMax(nums)
    let min = GetMin(nums)
    max - min
    
 
let Intro() =
    let lines = File.ReadAllLines(IntroFile)
    let mutable sum = 0
    for i = 0 to lines.Length - 1 do
        sum <- sum + ParseLine(lines[i])
    sum
  
let PartOne() =
    let lines = File.ReadAllLines(ActualFile)
    let mutable sum = 0
    for i = 0 to lines.Length - 1 do
        sum <- sum + ParseLine(lines[i])
    sum      
            
    
let Part2 =
    let lines = File.ReadAllLines(ActualFile)
    let answer = seq {
        for line in lines do
            let nums = ListToInt(line.Split('\t'))
            for x in nums do
            for y in nums do
                if x <> y && x % y  = 0 then  x / y else 0
    }
    answer |> Seq.sum


[<EntryPoint>]
let main(args) =
    Intro() |> printfn "%d"
    PartOne() |> printfn "%d"
    Part2 |> printfn "%d"
    0

File: AOC2017/Day2/CSharp/AOC2017-Day2/Program.cs
namespace AOC2017_Day2;

class Program {
    private const string IntroFile = @"D:\AdventOfCode\AOC2017\Day2\Inputs\intro.txt";
    private const string ActualFile = @"D:\AdventOfCode\AOC2017\Day2\Inputs\actual.txt";
    
    static void Main(string[] args)
    {
        Console.WriteLine($"Intro: {Intro()}");
        Console.WriteLine($"Part One: {PartOne()}");
        Console.WriteLine($"Part Two: {PartTwo()}");
    }

    static int Intro() {
        var sum = 0;
        var lines = File.ReadAllLines(IntroFile);
        foreach (var line in lines) {
            var numbers = line.Split('\t').Select(int.Parse).ToList();
            sum += numbers.Max() - numbers.Min();
        }

        return sum;
    }
    
    static int PartOne() {
        var sum = 0;
        var lines = File.ReadAllLines(ActualFile);
        foreach (var line in lines) {
            var numbers = line.Split('\t').Select(s => int.Parse(s.Trim())).ToList();
            sum += numbers.Max() - numbers.Min();
        }

        return sum;
    }
    
    static int PartTwo() {
        var sum = 0;
        var lines = File.ReadAllLines(ActualFile);
        foreach (var line in lines) {
            var numbers = line.Split('\t').Select(s => int.Parse(s.Trim())).ToList();
            for (var i = 0; i < numbers.Count; i++) {
                for (var x = i+1; x < numbers.Count; x++) {
                    if (numbers[i] % numbers[x] == 0) {
                        sum += numbers[i] / numbers[x];
                        Console.WriteLine($"{numbers[i]} / {numbers[x]}");
                        break;
                    }
                    if (numbers[x] % numbers[i] == 0) {
                        sum += numbers[x] / numbers[i];
                        Console.WriteLine($"{numbers[x]} / {numbers[i]}");
                        break;
                    }
                }
            }
        }
        
        return sum;
    }
}
